### PR TITLE
15 config parser

### PR DESCRIPTION
**Pull request by ioannnikulin**
Target Changes

- Modular parsing architecture: Split parsing logic into specialized modules - tokenization (ConfigParserTokens.cpp), server-level directives (ConfigParserElements.cpp), and location blocks (ConfigParserLocation.cpp)

- Token-based parsing: Implemented complete tokenization phase that separates file I/O from parsing logic, treating {, }, ; as individual tokens while grouping other characters into word tokens

- Comprehensive test suite: Added 5 unit tests covering basic server config, nested locations, multi-server setups, CGI handlers, and advanced features (all tests enabled and passing)

- Debug capability: Added printDebugInfo() helper in test suite for detailed configuration comparison during test failures

Note: The parser is currently tested only with valid configuration files. While it is designed to fail gracefully with most invalid input (throwing descriptive runtime errors), thorough error handling testing with malformed configurations, edge cases, and invalid values will be implemented in the next development phase.

Verification

- Basic configuration parsing: Run make test, verify 

1.   - testConfig0_BasicServer passes - confirms listen, server_name, root, and index directives are parsed correctly
2.   - Nested locations test: Check testConfig1_NestedLocations passes - validates multiple location blocks with different roots, methods, and upload settings work independently
3.   - Multi-server configuration: Verify testConfig2_MultiServer passes - ensures multiple server blocks can coexist with different ports (8080, 9090) and independent location configurations
4.   - CGI handler parsing: Confirm testConfig3_CgiHandler passes - validates multiple CGI extensions (.py, .php) map correctly to their executables
5.   - Advanced features test: Check testConfig4_Advanced passes - verifies error_page, upload, return redirects, and method restrictions all work together

- Debug output capability: Uncomment //printDebugInfo(expected, actual); in any test, run make test, observe detailed endpoint/route comparison showing interface, port, route paths, and equality checks at each level

Full validation suite: Run make test-github to execute all tests along with format checks, cppcheck, and clang-tidy - confirms parser integrates cleanly with existing codebase

### CODE REVIEW

File: sources/config_parser/ConfigParser.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ConfigParser::ConfigParser() : _index(0) {}
+ConfigParser::~ConfigParser() {}
+
+AppConfig ConfigParser::parse(const std::string& filename) {
+    std::string content = readFile(filename);
+    tokenize(content);
+    return buildConfigTree();
+}
+
+std::string ConfigParser::readFile(const std::string& filename) {
```
Comment (ioannnikulin) [inline sources/config_parser/ConfigParser.cpp]: (Not a blocker; style)
You can write `using std::string` once after includes, and you'll be able to save effort later, just writing `string`.

File: sources/config_parser/ConfigParser.cpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
+#include "ConfigParser.hpp"
+#include <unistd.h>   // for read, close
```
Comment (ioannnikulin) [inline sources/config_parser/ConfigParser.cpp]: If you need these comments to help you add includes in other files later, this is fine. On the other hand, if you're just tracking used and unused includes, there's no need for this, it's checked by some of the linters automatically.

**Review by shapkina-l — DISMISSED** [review]

File: sources/configuration/Endpoint.hpp [hunk 3]
```diff
@@ -1,22 +1,33 @@
...
-    static const std::string DEFAULT_INTERFACE;
-    static const int DEFAULT_PORT;
-
-    const std::string _interface;
-    const int _port;
+    std::string _interface;
+    int _port;
+    std::vector<std::string> _serverNames;
+    std::string _root;
+    size_t _maxRequestBodySizeBytes;
```
Comment (ioannnikulin) [inline sources/configuration/Endpoint.hpp]: please add a static constant default

File: sources/configuration/parser/ConfigParser.hpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    bool isEnd(const std::vector<string>& tokens, size_t index);
+
+public:
+    ConfigParser();
+    ~ConfigParser();
+
+    AppConfig parse(const std::string& filename);
+};
+
+struct LocationTempData {
```
Comment (ioannnikulin) [inline sources/configuration/parser/ConfigParser.hpp]: can we move the whole struct inside ConfigParser class and make it private to limit visibility scope?

File: linters/__pycache__/header_checker.cpython-312.pyc
Comment (ioannnikulin) [inline linters/__pycache__/header_checker.cpython-312.pyc]: please add to gitignore with some wildcards

File: sources/configuration/parser/ConfigParserLocation.cpp [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+    _index++;
+
+    if (isEnd(_tokens, _index) || _tokens[_index] != ";")
+        throw runtime_error("Missing ';' after index");
+
+    _index++;
+}
+
+static HttpMethodType stringToMethod(const string& s) {
```
Comment (ioannnikulin) [inline sources/configuration/parser/ConfigParserLocation.cpp]: please move to HttpMethodType.cpp and make public, I'm sure we'll need it again :)

File: tools/header_checker.py
Comment (ioannnikulin) [inline tools/header_checker.py]: we'd have to move this into target state header_checker... wonder how I forgot about it here. please don't delete it yet

File: Makefile [hunk 6]
```diff
@@ -28,6 +28,17 @@ APP_CONFIG_SRCS = $(addprefix $(SOURCE_F)/$(APP_CONFIG_F)/,$(APP_CONFIG_SRC_NAME
 
 # ------------------------------------------------------------
 
+CONFIG_PARSER_F = configuration/parser
```
Comment (ioannnikulin) [inline Makefile]: could you please rename to APP_CONFIG_PARSER_F? we might have other configurations

File: linters/__pycache__/header_checker.cpython-312.pyc
Comment (ioannnikulin) [reply linters/__pycache__/header_checker.cpython-312.pyc]: or maybe it's in dev already? please update the branch :)

File: Makefile [hunk 7]
```diff
@@ -28,6 +28,18 @@ APP_CONFIG_SRCS = $(addprefix $(SOURCE_F)/$(APP_CONFIG_F)/,$(APP_CONFIG_SRC_NAME
 
 # ------------------------------------------------------------
 
+APP_CONFIG_PARSER_F = configuration/parser
```
Comment (ioannnikulin) [inline Makefile]: please use $(APP_CONFIG_F)

File: sources/configuration/AppConfig.cpp [hunk 8]
```diff
@@ -42,6 +42,10 @@ bool AppConfig::operator==(const AppConfig& other) const {
     );
 }
 
+std::set<Endpoint>& AppConfig::getEndpoints() {
```
Comment (ioannnikulin) [inline sources/configuration/AppConfig.cpp]: please don't return as a reference to the original object, getters should return copies usually for original object safety. we should be able to change original fields ONLY through setters. please create a copy.

File: sources/configuration/CgiHandlerConfig.cpp [hunk 9]
```diff
@@ -1,19 +1,52 @@
...
-        _timeoutSeconds == other._timeoutSeconds
+        _extension == other._extension && _executablePath == other._executablePath &&
+        _rootPath == other._rootPath && _timeoutSeconds == other._timeoutSeconds
     );
 }
 
 CgiHandlerConfig::~CgiHandlerConfig() {
 }
+
+const string& CgiHandlerConfig::getExtension() const {
```
Comment (ioannnikulin) [inline sources/configuration/CgiHandlerConfig.cpp]: please return a copy

File: sources/configuration/Endpoint.cpp [hunk 10]
```diff
@@ -49,4 +82,40 @@ bool Endpoint::operator==(const Endpoint& other) const {
 
 Endpoint::~Endpoint() {
 }
+
+void Endpoint::setInterface(string interface) {
```
Comment (ioannnikulin) [inline sources/configuration/Endpoint.cpp]: please consider returning a reference to this to allow call chaining

File: sources/configuration/RouteConfig.cpp [hunk 11]
```diff
@@ -31,20 +37,73 @@ RouteConfig& RouteConfig::operator=(const RouteConfig& other) {
...
+
+    if (_redirections != other._redirections) {
+        return (false);
+    }
+
+    if (_folderConfigSection == NULL && other._folderConfigSection == NULL) {
+    } else if (_folderConfigSection == NULL || other._folderConfigSection == NULL) {
+        return (false);
+    } else {
+        if (!(*_folderConfigSection == *other._folderConfigSection)) {
```
Comment (ioannnikulin) [inline sources/configuration/RouteConfig.cpp]: ```if (((_folderConfigSection == NULL) != (other._folderConfigSection == NULL)) && (*_folderConfigSection != * other._folderConfigSection)) { return (false); )}```

File: sources/configuration/parser/ConfigParser.hpp [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+    static bool isEnd(const std::vector<std::string>& tokens, size_t index);
+
+public:
+    ConfigParser();
+    ConfigParser(const ConfigParser& other);
+    ConfigParser& operator=(const ConfigParser& other);
+    ~ConfigParser();
+
+    static const int BUFFER_SIZE = 4096;
```
[the remark this answers is not in the document]
Comment (shapkina-l) [reply sources/configuration/parser/ConfigParser.hpp]: we previously decided to move the readFile function into an anonymous namespace, outside of the ConfigParser class. because of that, readFile no longer has access to private members of ConfigParser, such as BUFFER_SIZE.

the same applies to POWER and the parseSizeValue method: since they are now used outside the class context, they also need to be public if we keep them in the anonymous namespace.

alternatively, we could move readFile and parseSizeValue back into the ConfigParser class, which would allow BUFFER_SIZE and POWER to remain private.

File: tools/header_checker.py
Comment (ioannnikulin) [inline tools/header_checker.py]: feel free to delete this file, I've taken the check into another PR

File: tests/config_files/advanced.conf [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+    error_page 404 /errors/404.html;
+    error_page 500 /errors/500.html;
+
+    location / {
+        root /srv/secure;
+        index index.html;
+    }
+
+    location /upload {
+        root /srv/uploads;
```
Comment (ioannnikulin) [inline tests/config_files/advanced.conf]: please add max body size

File: tests/config_files/advanced.conf [hunk 14]
```diff
@@ -0,0 +1,22 @@
+server {
+    listen 443;
+    server_name secure.example.com;
+
+    error_page 404 /errors/404.html;
+    error_page 500 /errors/500.html;
+
+    location / {
+        root /srv/secure;
+        index index.html;
```
Comment (ioannnikulin) [inline tests/config_files/advanced.conf]: Does the subject imply these are two different options?
◦ Enabling or disabling directory listing.
◦ Default file to serve when the requested resource is a directory

File: sources/configuration/RouteConfig.cpp [hunk 15]
```diff
@@ -1,50 +1,126 @@
 #include "RouteConfig.hpp"
 
 #include <cstddef>
+#include <string>
 
+#include "../http_methods/HttpMethodType.hpp"
```
Comment (ioannnikulin) [inline sources/configuration/RouteConfig.cpp]: please remove `../`
